src/lib/pathPlanner.ts

```diff
@@ -452,3 +455,74 @@ export function getSkillAnalysis<LU extends LearningUnit>({
 
 	return skillAnalyzedPath;
 }
+
+export function filterOutOfScopeLus<LU extends LearningUnit>(
+	goal: Skill[],
+	learningUnits: ReadonlyArray<LU>,
+	knowledge: Skill[]
+): LU[] {
+	const filteredLearningUnits = learningUnits.filter(lu =>
+		lu.teachingGoals.some(skill => goal.includes(skill))
+	);
+
+	goal.forEach(skill => {
+		skill.nestedSkills.forEach(nestedSkill => {
+			const lus = learningUnits.filter(learningUnit =>
+				learningUnit.teachingGoals.some(skill => skill.id == nestedSkill)
+			);
+			const luNotExist = lus.filter(lu => !filteredLearningUnits.includes(lu));
+			filteredLearningUnits.push(...luNotExist);
+		});
+	});
+
+	const inScopeLearningUnits = filteredLearningUnits.slice();
+
+	while (filteredLearningUnits.length > 0) {
+		const learningUnit = filteredLearningUnits.pop();
+		if (
+			learningUnit.requiredSkills.length == 0 ||
+			learningUnit.requiredSkills.every(skill => knowledge.includes(skill))
+		) {
+			continue;
+		}
+		let requiredLearningUnit = learningUnits.filter(lu =>
+			lu.teachingGoals.some(skill => learningUnit.requiredSkills.includes(skill))
+		);
+
+		learningUnit.requiredSkills.forEach(skill => {
+			requiredLearningUnit.push(
```

Comment:
Is it possible to make it more readable by avoding this side-effect or by spliting it into 2 statements?

For instance:
```typescript
const predecessorLUs = learningUnit.requiredSkills.forEach(skill => {
    learningUnits.filter(...)
                 .filter(...)
});
requiredLearningUnit.push(predecessorLUs);
```